File: app/src/main/java/com/funny/translation/translation/TranslationBV2AV.java
package com.funny.translation.translation;

import android.text.TextUtils;

import com.funny.translation.bean.Consts;
import com.funny.translation.utils.FunnyBvToAv;
import com.funny.translation.utils.StringUtil;

public class TranslationBV2AV extends BasicTranslationTask{

    public TranslationBV2AV(String sourceString, short sourceLanguage, short targetLanguage) {
        super(sourceString, sourceLanguage, targetLanguage);
    }

    @Override
    public String getBasicText(String url) throws TranslationException {
        String result="";
        String inputText = sourceString;
        if (TextUtils.isEmpty(inputText))
        {
            return result;
        }
        if (StringUtil.isNumber(inputText))
        {
            result = FunnyBvToAv.enc(Long.parseLong(inputText));
        }else{
            long av=0;
            if ((av = StringUtil.findAv(inputText)) > 0){
                result = FunnyBvToAv.enc(av);
            }else{
                String bv="";
                if (!(bv = StringUtil.findBv(inputText)).equals("")){
                    result = FunnyBvToAv.dec(bv);
                }
            }
        }
        return result;
    }

    @Override
    public TranslationResult getFormattedResult(String basicText) throws TranslationException {
        TranslationResult result = new TranslationResult(getEngineKind());
        if (!basicText.equals("")){
            result.setBasicResult(basicText);
        }else{
            throw new TranslationException(Consts.ERROR_NO_BV_OR_AV);
        }
        return result;
    }

    @Override
    public String madeURL() {
        return null;
    }

    @Override
    public boolean isOffline() {
        return true;
    }

    @Override
    public short getEngineKind() {
        return Consts.ENGINE_BV_TO_AV;
    }
}

File: app/src/main/java/com/funny/translation/translation/TranslationBV2AV.kt
package com.funny.translation.translation

import android.text.TextUtils
import com.funny.translation.bean.Consts
import com.funny.translation.trans.TranslationException
import com.funny.translation.trans.TranslationResult
import com.funny.translation.utils.FunnyBvToAv
import com.funny.translation.utils.StringUtil

class TranslationBV2AV(sourceString: String?, sourceLanguage: Short, targetLanguage: Short) :
    BasicTranslationTask(
        sourceString!!, sourceLanguage, targetLanguage
    ) {
    @Throws(TranslationException::class)
    override fun getBasicText(url: String): String {
        var result = ""
        val inputText = sourceString
        if (TextUtils.isEmpty(inputText)) {
            return result
        }
        if (StringUtil.isNumber(inputText)) {
            result = FunnyBvToAv.enc(inputText.toLong())
        } else {
            var av: Long = 0
            if (StringUtil.findAv(inputText).also { av = it } > 0) {
                result = FunnyBvToAv.enc(av)
            } else {
                var bv: String? = ""
                if (StringUtil.findBv(inputText).also { bv = it } != "") {
                    result = FunnyBvToAv.dec(bv)
                }
            }
        }
        return result
    }

    @Throws(TranslationException::class)
    override fun getFormattedResult(basicText: String): TranslationResult {
        val result = TranslationResult(engineKind)
        if (basicText != "") {
            result.setBasicResult(basicText)
        } else {
            throw TranslationException(Consts.ERROR_NO_BV_OR_AV)
        }
        return result
    }

    override fun madeURL(): String {
        return ""
    }

    override val isOffline: Boolean
        get() = true
    override val engineKind: Short
        get() = Consts.ENGINE_BV_TO_AV
}